File: cpp/leetcode/editor/cn/[167]两数之和 II - 输入有序数组.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>
#include <utility>
#include <memory>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    vector<int> twoSum(vector<int> &numbers, int target) {
        auto left = numbers.begin();
        auto right = numbers.end() - 1;
        while (left < right) {
            auto sum = *left + *right;
            if (sum == target) {
                return {static_cast<int>(distance(numbers.begin(), left) + 1),
                        static_cast<int>(distance(numbers.begin(), right) + 1)};
            } else if (sum < target) {
                left++;
            } else if (sum > target) {
                right--;
            }
        }
        return {-1, -1};
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}

File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[167]两数之和 II - 输入有序数组.scala
package leetcode.editor.cn
object TwoSumIiInputArrayIsSorted {
//leetcode submit region begin(Prohibit modification and deletion)
  object Solution {
    def twoSum(numbers: Array[Int], target: Int): Array[Int] = {
      @annotation.tailrec
      def impl(left: Int, right: Int): Array[Int] = {
        (left, right) match {
          case (l, r) if l >= r => Array(-1, -1)
          case (l, r) if numbers(l) + numbers(r) == target =>
            Array(l + 1, r + 1)
          case (l, r) if numbers(l) + numbers(r) < target => impl(l + 1, r)
          case (l, r)                                     => impl(l, r - 1)
        }
      }

      impl(0, numbers.length - 1)

    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}
